components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -421,121 +418,98 @@ export class GitpodServerImpl<Client extends GitpodClient, Server extends Gitpod
+        // no matter if the workspace is shared or not, you cannot create a new instance
+        await this.guardAccess({ kind: "workspaceInstance", subject: undefined, workspace }, "create");
 
-            if (workspace.type !== "regular") {
-                throw new ResponseError(ErrorCodes.PERMISSION_DENIED, "Cannot (re-)start irregular workspace.");
-            }
+        if (workspace.type !== "regular") {
+            throw new ResponseError(ErrorCodes.PERMISSION_DENIED, "Cannot (re-)start irregular workspace.");
+        }
 
-            if (workspace.deleted) {
-                throw new ResponseError(ErrorCodes.PERMISSION_DENIED, "Cannot (re-)start a deleted workspace.");
-            }
-            const envVars = this.userDB.getEnvVars(user.id);
+        if (workspace.deleted) {
+            throw new ResponseError(ErrorCodes.PERMISSION_DENIED, "Cannot (re-)start a deleted workspace.");
+        }
+        const envVars = this.userDB.getEnvVars(user.id);
 
-            await mayStartPromise;
+        await mayStartPromise;
 
-            // at this point we're about to actually start a new workspace
-            return await this.workspaceStarter.startWorkspace({ span }, workspace, user, await envVars, {
-                forceDefaultImage: !!options.forceDefaultImage
-            });
-        } catch (e) {
-            TraceContext.logError({ span }, e);
-            throw e;
-        } finally {
-            span.finish();
-        }
+        // at this point we're about to actually start a new workspace
+        return await this.workspaceStarter.startWorkspace(ctx, workspace, user, await envVars, {
+            forceDefaultImage: !!options.forceDefaultImage
+        });
     }
 
-    public async stopWorkspace(workspaceId: string): Promise<void> {
+    public async stopWorkspace(ctx: TraceContext, workspaceId: string): Promise<void> {
+        ctx.span?.setTag("workspaceId", workspaceId);
+
         const user = this.checkUser('stopWorkspace', undefined, { workspaceId });
         const logCtx = { userId: user.id, workspaceId };
 
-        const span = opentracing.globalTracer().startSpan("stopWorkspace");
-        span.setTag("workspaceId", workspaceId);
-
-        try {
-            const workspace = await this.internalGetWorkspace(workspaceId, this.workspaceDb.trace({ span }));
-            if (workspace.type === 'prebuild') {
-                // If this is a team prebuild, any team member can stop it.
-                const teamMembers = await this.getTeamMembersByProject(workspace.projectId);
-                await this.guardAccess({ kind: "workspace", subject: workspace, teamMembers }, "get");
-            } else {
-                // If this is not a prebuild, or it's a personal prebuild, only the workspace owner can stop it.
-                await this.guardAccess({ kind: "workspace", subject: workspace }, "get");
-            }
-
-            this.internalStopWorkspace({ span }, workspace).catch(err => {
```

Comment:
nice to see that pattern gone (here)

Reply:
Yep. In some places we still have sth similar for local spans. We could come up with a generic, type-safe interceptor for functions as well... but one step at a time. :wink:

Reply:
ok with me to leave it for now. it's a good trade after all.